File: java/src/main/java/cn/az/code/year2021/may/CountPrimes.java
package cn.az.code.year2021.may;

/**
 * @author az
 * @since 2021-05-10
 */
public class CountPrimes {

    // https://leetcode.com/problems/count-primes/discuss/57588/My-simple-Java-solution

    public int countPrimes(int n) {
      if(n <=1 ) return 0;
    
    boolean[] notPrime = new boolean[n];        
    notPrime[0] = true; 
    notPrime[1] = true; 

    for(int i = 2; i < Math.sqrt(n); i++){
        if(!notPrime[i]){
            for(int j = 2; j*i < n; j++){
                notPrime[i*j] = true; 
            }
        }
    }
    
    int count = 0; 
    for(int i = 2; i< notPrime.length; i++){
        if(!notPrime[i]) count++;
    }
    return count; 
    }
}

File: code/cpp/2021/may/count-primes.cc
#include <vector>

using namespace std;

class Solution
{
public:
    // TLE solution BF
    int countPrimes(int n)
    {
        int count = 0;
        for (int i = 0; i <= n; ++i)
        {
            if (isPrime(i))
            {
                count++;
            }
        }
        return count;
    }

    bool isPrime(int n)
    {
        if (n == 0 || n == 1)
        {
            return false;
        }
        for (int i = 2; i <= n / 2; ++i)
        {
            if (n % i == 0)
            {
                return false;
            }
        }
        return true;
    }

private:
    int _countPrimes(int n)
    {
        int count = 0;
        vector<bool> sieve(n, true);
        for (int i = 2; i < n; i++)
        {
            if (sieve[i])
            {
                count++;
                int k = 2;
                while (k * i < n)
                {
                    sieve[k * i] = false;
                    k++;
                }
            }
        }
        return count;
    }
};